File: packages/core/src/schedule.js
// Weekly schedule windows. A rule = { days: [1..7], start: minutes, end: minutes }.
// `days` are the days a window STARTS on. end <= start means the window crosses
// midnight (end === start means a full 24 hours).
import { addDays, atMinutes, isoWeekday, startOfDay } from './time.js'

function durationMinutes(rule) {
  const d = rule.end - rule.start
  return d > 0 ? d : d + 1440
}

/** Window instance that starts on the local day `dayMs`, or null if not scheduled. */
function windowStartingOn(rule, dayMs) {
  if (!rule.days.includes(isoWeekday(dayMs))) return null
  const start = atMinutes(dayMs, rule.start)
  const endDay = rule.end > rule.start ? dayMs : addDays(startOfDay(dayMs), 1)
  const end = atMinutes(endDay, rule.end)
  return { start, end }
}

/** The window containing `now`, or null. */
export function windowAt(rule, now) {
  const today = startOfDay(now)
  for (const day of [today, addDays(today, -1)]) {
    const w = windowStartingOn(rule, day)
    if (w && now >= w.start && now < w.end) return w
  }
  return null
}

/** Most recent window whose end is <= now. */
export function previousWindow(rule, now) {
  const today = startOfDay(now)
  for (let i = 0; i <= 8; i++) {
    const w = windowStartingOn(rule, addDays(today, -i))
    if (w && w.end <= now) return w
  }
  return null
}

/** Start of the next window strictly after now, or null if the rule has no days. */
export function nextWindowStart(rule, now) {
  const today = startOfDay(now)
  for (let i = 0; i <= 8; i++) {
    const w = windowStartingOn(rule, addDays(today, i))
    if (w && w.start > now) return w.start
  }
  return null
}

const WEEK = 7 * 1440

/** Intervals in "minutes since Monday 00:00", wrapped into [0, WEEK). */
function weekIntervals(rule) {
  const out = []
  for (const day of rule.days) {
    const s = (day - 1) * 1440 + rule.start
    const e = s + durationMinutes(rule)
    if (e <= WEEK) out.push([s, e])
    else {
      out.push([s, WEEK])
      out.push([0, e - WEEK])
    }
  }
  return out
}

/** True if the two weekly schedules share any minute. */
export function rulesOverlap(a, b) {
  const ia = weekIntervals(a)
  const ib = weekIntervals(b)
  return ia.some(([s1, e1]) => ib.some(([s2, e2]) => s1 < e2 && s2 < e1))
}

export function validateSchedule(rule) {
  const errors = []
  if (!Array.isArray(rule.days) || rule.days.length === 0) errors.push('Pick at least one day.')
  else if (!rule.days.every((d) => Number.isInteger(d) && d >= 1 && d <= 7)) errors.push('Invalid day.')
  for (const k of ['start', 'end']) {
    if (!Number.isInteger(rule[k]) || rule[k] < 0 || rule[k] > 1439) errors.push(`Invalid ${k} time.`)
  }
  return errors
}

File: agent/internal/core/schedule.go
package core

import (
	"math"
	"time"
)

// Local-clock helpers. Regimen reads the system clock and stores no time
// zone (SPEC.md "Timezone Handling"). All instants are epoch milliseconds.

func local(ms int64) time.Time { return time.UnixMilli(ms).In(time.Local) }

// IsoWeekday is 1 = Monday ... 7 = Sunday.
func IsoWeekday(ms int64) int {
	d := int(local(ms).Weekday())
	if d == 0 {
		return 7
	}
	return d
}

// StartOfDay is local midnight of the day containing ms.
func StartOfDay(ms int64) int64 {
	t := local(ms)
	return time.Date(t.Year(), t.Month(), t.Day(), 0, 0, 0, 0, time.Local).UnixMilli()
}

// AddDays adds calendar days while keeping the local wall-clock time (DST safe).
func AddDays(ms int64, n int) int64 {
	t := local(ms)
	return time.Date(t.Year(), t.Month(), t.Day()+n, t.Hour(), t.Minute(), t.Second(), t.Nanosecond(), time.Local).UnixMilli()
}

// AtMinutes is local midnight of dayMs plus minutes of wall-clock time.
func AtMinutes(dayMs int64, minutes int64) int64 {
	t := local(StartOfDay(dayMs))
	h := int64(math.Floor(float64(minutes) / 60))
	m := minutes % 60 // same sign rules as JavaScript's %
	return time.Date(t.Year(), t.Month(), t.Day(), int(h), int(m), 0, 0, time.Local).UnixMilli()
}

// Window is one instance of a weekly schedule, [Start, End) in epoch ms.
type Window struct {
	Start int64
	End   int64
}

func ruleMinutes(rule any, k string) int64 { return int64(Num(Get(rule, k))) }

func includesDay(rule any, day int) bool {
	for _, d := range Arr(Get(rule, "days")) {
		if StrictEq(d, float64(day)) {
			return true
		}
	}
	return false
}

// windowStartingOn returns the window instance that starts on the local day dayMs.
func windowStartingOn(rule any, dayMs int64) *Window {
	if !includesDay(rule, IsoWeekday(dayMs)) {
		return nil
	}
	start := AtMinutes(dayMs, ruleMinutes(rule, "start"))
	endDay := dayMs
	if !(Num(Get(rule, "end")) > Num(Get(rule, "start"))) {
		endDay = AddDays(StartOfDay(dayMs), 1)
	}
	end := AtMinutes(endDay, ruleMinutes(rule, "end"))
	return &Window{Start: start, End: end}
}

// WindowAt is the window containing now, or nil.
func WindowAt(rule any, now int64) *Window {
	today := StartOfDay(now)
	for _, day := range []int64{today, AddDays(today, -1)} {
		if w := windowStartingOn(rule, day); w != nil && now >= w.Start && now < w.End {
			return w
		}
	}
	return nil
}

// PreviousWindow is the most recent window whose end is <= now.
func PreviousWindow(rule any, now int64) *Window {
	today := StartOfDay(now)
	for i := 0; i <= 8; i++ {
		if w := windowStartingOn(rule, AddDays(today, -i)); w != nil && w.End <= now {
			return w
		}
	}
	return nil
}
